cpp/ql/src/experimental/Security/CWE/CWE-783/OperatorPrecedenceLogicErrorWhenUseBitwiseOrLogicalOperations.ql

```diff
@@ -0,0 +1,179 @@
+  lowerBound(exp) != -8388608 and
+  lowerBound(exp) != -65536 and
+  lowerBound(exp) != -32768 and
+  lowerBound(exp) != -128
+  or
+  lowerBound(exp) = 0 and
+  upperBound(exp) = 1
+}
+
+/** Holds if expressions are of different size or range */
+predicate isDifferentSize(Expr exp1, Expr exp2, Expr exp3) {
+  exp1.getType().getSize() = exp2.getType().getSize() and
+  exp1.getType().getSize() != exp3.getType().getSize()
+  or
+  (
+    isRealRange(exp1) and
+    isRealRange(exp2) and
+    isRealRange(exp3)
+  ) and
+  upperBound(exp1).maximum(upperBound(exp2)) - upperBound(exp1).minimum(upperBound(exp2)) < 16 and
+  lowerBound(exp1).maximum(lowerBound(exp2)) - lowerBound(exp1).minimum(lowerBound(exp2)) < 16 and
+  (
+    upperBound(exp1).maximum(upperBound(exp3)) - upperBound(exp1).minimum(upperBound(exp3)) > 256 or
+    lowerBound(exp1).maximum(lowerBound(exp2)) - lowerBound(exp1).minimum(lowerBound(exp2)) > 256
+  )
+}
+
+/** Holds if it is possible to get different values of the expression */
+predicate isDifferentResults(
+  Expr exp1, Expr exp2, Expr exp3, BinaryBitwiseOperation op1, BinaryBitwiseOperation op2
+) {
+  (
+    isRealRange(exp1) and
+    isRealRange(exp2) and
+    isRealRange(exp3)
+  ) and
+  exists(int i1, int i2, int i3 |
+    i1 in [lowerBound(exp1).floor() .. upperBound(exp1).floor()] and
+    i2 in [lowerBound(exp2).floor() .. upperBound(exp2).floor()] and
+    i3 in [lowerBound(exp3).floor() .. upperBound(exp3).floor()] and
+    (
+      op1 instanceof BitwiseOrExpr and
+      op2 instanceof BitwiseAndExpr and
+      i1.bitOr(i2).bitAnd(i3) != i2.bitAnd(i3).bitOr(i1)
+      or
+      op1 instanceof BitwiseOrExpr and
+      op2 instanceof BitwiseXorExpr and
+      i1.bitOr(i2).bitXor(i3) != i2.bitXor(i3).bitOr(i1)
+      or
+      op1 instanceof BitwiseXorExpr and
+      op2 instanceof BitwiseAndExpr and
+      i1.bitXor(i2).bitAnd(i3) != i2.bitAnd(i3).bitXor(i1)
+    )
+  )
+}
+
+from Expr exp, string msg
+where
+  isLogicalOrAndExpr(exp) and
+  msg = "Logical and has a higher priority."
+  or
+  isLogicalOrandBitwise(exp) and
+  msg = "Binary operations has higher priority."
+  or
+  isBitwiseandBitwise(exp) and
+  isDifferentSize(exp.(BinaryBitwiseOperation).getLeftOperand(),
+    exp.(BinaryBitwiseOperation).getRightOperand().(BinaryBitwiseOperation).getLeftOperand(),
+    exp.(BinaryBitwiseOperation).getRightOperand().(BinaryBitwiseOperation).getRightOperand()) and
+  msg = "Expression ranges do not match operation precedence."
+  or
+  isBitwiseandBitwise(exp) and
+  isDifferentResults(exp.(BinaryBitwiseOperation).getLeftOperand(),
+    exp.(BinaryBitwiseOperation).getRightOperand().(BinaryBitwiseOperation).getLeftOperand(),
+    exp.(BinaryBitwiseOperation).getRightOperand().(BinaryBitwiseOperation).getRightOperand(),
+    exp.(BinaryBitwiseOperation),
+    exp.(BinaryBitwiseOperation).getRightOperand().(BinaryBitwiseOperation)) and
+  msg = "specify the priority with parentheses."
```

Comment:
Similarly here: Could you add a comment that explains what kind of error this catches?

Reply:
according to the original idea, this predicate was supposed to single out those expressions that, as a result of identifying the priority with parentheses, will give different values.
as a consequence, this piece of code was supposed to find errors associated with possible outcomes of operations.

Reply:
I see. I think it will be helpful if you could write that in the code. Maybe as a comment before the `isBitwiseandBitwise(exp) and` line?